Added scrollEnabled, oneFingerScrollEnabled, twoFingersScrollEnabled and threeFingersScrollEnabled

RayyanTahir opened the pull request:
This PR adds:

1. `scrollEnabled` if set to false, scrolling will be disabled. This is set to true by default
2. `oneFingerScrollEnabled` if set to false, scrolling cannot be done with one finger. This is set to true by default
3. `twoFingersScrollEnabled` if set to false, scrolling cannot be done with two fingers. This is set to true by default
4. `threeFingersScrollEnabled` if set to false, scrolling cannot be done with three fingers. This is set to true by default

Fixes/Satisfies feature request:
#101

Commit:
Added scrollEnabled and singleFingerScrollEnabled

Commit:
Added scrollEnabled and singleFingerScrollEnabled for ZoomSurfaceView

RayyanTahir commented:
@natario1 Please review this when you have time. Thanks

markusressel commented:
Isn't the naming convention a bit confusing when we already have `horizontalPanEnabled` and `horizontalPanEnabled`?

markusressel reviewed (changes requested):
The code looks fine to me, I only have an issue with the naming of those attributes.
Since @natario1 most likely want's to have a stable API we can't rename existing attributes so I would suggest to call yours `...Pan...` instead of `...Scroll...`.

And: those attributes have to be added to the documentation too.

natario1 commented on library/src/main/java/com/otaliastudios/zoom/ZoomApi.kt:
```diff
@@ -229,6 +229,20 @@ interface ZoomApi {
+     *
+     * @param enabled true enables scroll gesture, false disables it
+     */
+    fun setScrollEnabled(enabled: Boolean)
```
Can you add these two in README?

RayyanTahir replied:
I'll add it to the README

natario1 commented on library/src/main/java/com/otaliastudios/zoom/internal/gestures/ScrollFlingDetector.kt:
```diff
@@ -144,6 +146,8 @@ internal class ScrollFlingDetector(
             @ZoomApi.ScaledPan distanceY: Float
     ): Boolean {
+        if (!scrollEnabled) return false
+        if (!singleFingerScrollEnabled && e1 != null && e1.pointerCount == 1) return false
```
You should use e2 here, as e1 is the starting event and should always have 1 pointer.

Could you write it like this to make it more readable:

val isSingleFinger = e2?.pointerCount == 1
if (!singleFingerScrollEnabled && isSingleFinger) return false

RayyanTahir replied:
Sure, I'll fix it

natario1 reviewed (changes requested):
I think naming is OK! We want to have both engine movement control (zoomEnabled, panEnabled) and selective control over the gestures that create these movements (flingEnabled, scrollEnabled, ... ).

Now that I see this, it would be cool to also have `doubleFingerScrollEnabled` and `tripleFingerScrollEnabled`, if you have time. If not, please just address the comments above. Thanks!

RayyanTahir commented:
Thankyou for your feedbacks @natario1 @markusressel. If the naming convention is alright, I'll move forward with double and triple finger scroll properties along with the changes required in README and motion event handling

natario1 commented:
Maybe oneFingerScroll, twoFingersScroll and threeFingersScroll are better than single, double and triple. I would do this way

natario1 commented:
I just read your first comment @markusressel I had missed it.

I don't see this as redundant, pan is an engine movement that can be generated by a few gestures - flinging, single finger scroll, two fingers scroll, three fingers scroll, and even while pinching. We're giving selective control here.

For instance, we also already have flingEnabled. In a future release we can use the prefix "gesture" for all these APIs (or "movement" suffix for pan and zoom instead, like panMovementEnabled), but for now I think docs will be enough, what do you think?

This is reflected in code, as these flags go to gesture detectors, while movement flags go to movement managers.

markusressel commented:
@natario1 sounds good to me :+1:

natario1 commented:
Cool! When this is merged I will add the missing flags, which are scrollWhilePinchingEnabled and pinchEnabled. pinchEnabled is quite redundant with zoomEnabled at this point but maybe in the future we can implement other ways to zoom. For example google maps lets you zoom with a single finger (tap + second tap and hold)

So @RayyanTahir , you can go on!

Commit:
Added oneFingerScrollEnabled, twoFingersScrollEnabled and threeFingersScrollEnabled flags, replacing singleFingerScrollEnabled flag
Added details to README for the added flags
Fixed motion event handling for the added flags

RayyanTahir commented:
Thankyou, @natario1 @markusressel please review my changes. 

1. I have added details of flags to README
2. Fixed motion event handling to detect pointer count using e2
3. Used the new naming convention; oneFingerScrollEnabled, twoFingersScrollEnabled and threeFingersScrollEnabled

markusressel commented on library/src/main/java/com/otaliastudios/zoom/ZoomApi.kt:
```diff
@@ -237,11 +237,25 @@ interface ZoomApi {
+    /**
+     * Controls whether two fingers scroll gesture is enabled or not.
+     *
+     * @param enabled true enables twi fingers scroll gesture, false disables it
```
typo: **two**

RayyanTahir replied:
🤦‍♂️ my bad. Fixing

markusressel commented on library/src/main/res/values/attrs.xml:
```diff
@@ -10,7 +10,9 @@
         <attr name="scrollEnabled" format="boolean" />
-        <attr name="singleFingerScrollEnabled" format="boolean" />
+        <attr name="oneFingerScrollEnabled" format="boolean" />
+        <attr name="twoFingersScrollEnabled" format="boolean" />
```
do you really want to go with finger vs finger**s**?
`twoFingerScrollEnabled` would be ok in my brain and it would be more consistent only changing the count.

RayyanTahir replied:
While it would make sense to go with `twoFingerScrollEnabled`, I think we should still go with `twoFingersScrollEnabled` since @natario1 specifically suggested here https://github.com/natario1/ZoomLayout/pull/102#issuecomment-478974434

Commit:
Typo fix

RayyanTahir commented:
@natario1 @markusressel Is everything good to go for merge?

markusressel commented:
I want to wait for @natario1 opinion on the naming subject because if we want to change this later it would be a breaking change. Other than that its fine.

natario1 commented:
It's fine for me! Name seems OK... Thanks for your work @RayyanTahir !
